### PR TITLE
[wip] Superglue: better glue-code performance

### PR DESCRIPTION
**Pull request by apendleton**
The goal of this branch is to augment the functionality and, especially, the performance, of the glue code introduced in #10 . The gameplan:
- [x] Introduce a framework for restricting the words that we index for fuzzy-searching (at index time) and attempt to fuzzy-lookup (at query time) -- see #29 . There are two parts here:
    - [x] don't try to fuzzy-correct any words that contain digits or `#`
    - [x] introduce a mechanism for limiting the eligible scripts for fuzzy-indexing/matching, and only index/match those scripts. For this work I poached and slightly modified https://github.com/servo/unicode-script/blob/master/src/tables.rs -- modifications are to switch to a newer version of the underlying Unicode tables using their supplied scripts, and to make the generated Rust tables public, which they aren't in the upstream library. There are some corresponding LICENSE changes here, per consultation with Legal.
- [x] Be smarter about the recursive enumeration of possible variant combinations, to actually walk the phrase graph while recursing and prune failed search avenues. This is an alternative to the FST `Automaton` approach @boblannon is pursuing, so might get ripped out again if that proves faster.
- [x] TBD profiler-driven micro optimizations

### CODE REVIEW

**Comment by apendleton** [comment]
Benches so far on this branch, mostly improved due to no longer fuzzy-correcting numbers:
```
Running target/release/deps/benchmarks-24c6206cf22e98ea
Gnuplot not found, disabling plotting
glue/fuzzy_match        time:   [141.32 us 147.39 us 156.02 us]
                   change: [-37.413% -33.889% -28.860%] (p = 0.00 < 0.05)
                   Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
3 (3.00%) high mild
3 (3.00%) high severe
glue/fuzzy_match_prefix time:   [87.560 us 90.463 us 93.412 us]
                   change: [-57.330% -55.540% -53.713%] (p = 0.00 < 0.05)
                   Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
4 (4.00%) high mild
```

File: src/phrase/mod.rs [hunk 1]
```diff
@@ -74,34 +74,14 @@ impl PhraseSet {
     }
 
     // TODO: this needs to get called inside contains_prefix when final word is QueryWord::prefix <15-05-18, boblannon> //
```
Comment (boblannon) [inline src/phrase/mod.rs]: this TODO can go away

File: src/phrase/mod.rs [hunk 2]
```diff
@@ -43,18 +43,18 @@ impl PhraseSet {
...
-            None => return Ok(false),
-            Some(..) => return Ok(true),
+            None => Ok(false),
+            Some(addr) => {
+                if phrase.has_prefix {
+                    Ok(self.matches_prefix_range(addr, phrase.prefix_key_range().unwrap()))
+                } else {
+                    Ok(true)
+                }
+            }
```
Comment (boblannon) [inline src/phrase/mod.rs]: looks good to me!

File: src/phrase/mod.rs [hunk 3]
```diff
@@ -74,34 +74,14 @@ impl PhraseSet {
     }
 
     // TODO: this needs to get called inside contains_prefix when final word is QueryWord::prefix <15-05-18, boblannon> //
-    fn contains_prefix_with_range(&self, phrase: QueryPhrase) -> bool {
-        let (sought_min_key, sought_max_key) = phrase.prefix_key_range().unwrap();
+    fn matches_prefix_range(&self, start_position: CompiledAddr, key_range: (Vec<u8>, Vec<u8>)) -> bool {
+        let (sought_min_key, sought_max_key) = key_range;
```
Comment (boblannon) [inline src/phrase/mod.rs]: yeah, this will be super helpful for the automaton approach.

**Comment by apendleton** [comment]
Just flamegraphed the current state of this branch. Have only dug into the exact, not prefix, so far, but the conclusions are unexpected. Mostly: The recursive-strategy version is fast enough now that it's not a significant contributor to the overall runtime. We currently spent 7x as much time fuzzy-matching words as we do on phrase graph traversal -- fuzzy lookup currently dominates the total search time. Further, well more than half of *that* is spent in the `damerau_levenshtein` confirmation step. In other words, 60% of the whole glue whole-phrase lookup is just spent on Damerau-Levenshtein. Oof.

This suggests to me that a bunch more digging is needed on that front. Specifically:
* are there faster DL implementations out there, maybe using SIMD or something else fancy?
* are there special cases for DL computation (e.g., shortcuts that can be used on very short words?)
* are there approaches for doing a whole bunch of DL comparisons (like, comparing each of A1...Ak to B) all at once in some way that saves work?

**Comment by boblannon** [comment]
> Goal is to get another 2-3x improvement from here.

...is that work for another PR, or is this one ready to review/merge?

**Comment by apendleton** [comment]
@boblannon I was thinking we would choose a strategy for doing the multi-search before I merged this one, but I can merge it sooner if it would make comparisons between the two easier (also, being honest: I found the numbers-oriented tests in the phrase code to be a little intimidating to reason about, so I was going to steal yours if you were writing them for the automaton 😆 ).

I can merge sooner, but the code I've written so far probably needs better tests and comments before I do so. I'm confident that the code works correctly as written, but from fuzz-testing, not from unit tests.

**Comment by boblannon** [comment]
okay, yeah i have some unit tests in https://github.com/mapbox/fuzzy-phrase/pull/34. They're very basic, though. I won't be surprised if they're not covering all potential problem cases.

I think I want to move on to benchmarking via glue, though, and I was trying to get a sense of whether master would change in the short term.  It sounds like this is at a point where it would be preferred over master.

**Comment by apendleton** [comment]
Cool. I think, then, that I'll hold on writing more tests for now (we can wait and do it once we know more about what specifically we're going to do), but I'll add some more comments to explain what some of the new stuff does. The Levenshtein thing, in particular, is a doozy.

**Pull request merged**